File: internal/domain/rpc/loyalty/loyalty.go
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

package loyaltyrpc

import "ItsBagelBot/internal/domain/rpc"

type Request struct {
	UserID      string `json:"user_id"`
	ViewerID    string `json:"viewer_id,omitempty"`
	ViewerLogin string `json:"viewer_login,omitempty"`
	Name        string `json:"name,omitempty"`
	NewName     string `json:"new_name,omitempty"`
	Scope       string `json:"scope,omitempty"`
	Command     string `json:"command,omitempty"`
	Value       int64  `json:"value,omitempty"`
	Limit       int    `json:"limit,omitempty"`
}

type Balance struct {
	ViewerID     string `json:"viewer_id"`
	ViewerLogin  string `json:"viewer_login,omitempty"`
	ViewerName   string `json:"viewer_name,omitempty"`
	Points       int64  `json:"points"`
	WatchSeconds uint64 `json:"watch_seconds"`
}

type Counter struct {
	Name  string `json:"name"`
	Scope string `json:"scope"`
	Value int64  `json:"value"`
}

type CounterEntry struct {
	ViewerID    string `json:"viewer_id"`
	ViewerLogin string `json:"viewer_login,omitempty"`
	ViewerName  string `json:"viewer_name,omitempty"`
	Command     string `json:"command,omitempty"`
	Value       int64  `json:"value"`
}

type CounterRank struct {
	UserID string `json:"user_id"`
	Value  int64  `json:"value"`
}

type Reply struct {
	Balance       *Balance       `json:"balance,omitempty"`
	TargetBalance *Balance       `json:"target_balance,omitempty"`
	Top           []Balance      `json:"top,omitempty"`
	Counter       *Counter       `json:"counter,omitempty"`
	Counters      []Counter      `json:"counters,omitempty"`
	Entries       []CounterEntry `json:"entries,omitempty"`
	Board         []CounterRank  `json:"board,omitempty"`
	Found         bool           `json:"found,omitempty"`
	Spent         bool           `json:"spent,omitempty"`
	rpc.Refusal
}

File: console/shared/lib/loyalty.ts
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

// --- Loyalty ----------------------------------------------------------------
// The loyalty economy: viewers earn points from subs, resubs, gift subs,
// cheers and watch time (a 5-minute tick over everyone in chat while live).
// Rates live in the "loyalty" module blob; standings and counters live in the
// loyalty service (bagel.rpc.loyalty.*).

// LoyaltyConfig mirrors sesame's LoyaltyModuleConfig blob: 0 means "use the
// default", a negative value switches that source off.
export interface LoyaltyConfig {
  pointsName: string;
  subPoints: number;
  resubPoints: number;
  giftSubPoints: number;
  cheerPointsPer100: number;
  watchPointsPerTick: number;
}

// LOYALTY_DEFAULTS are the effective rates behind a zero value, mirrored from
// sesame so the form can show what "default" means.
export const LOYALTY_DEFAULTS: LoyaltyConfig = {
  pointsName: 'points',
  subPoints: 500,
  resubPoints: 500,
  giftSubPoints: 100,
  cheerPointsPer100: 50,
  watchPointsPerTick: 10
};

export function blankLoyaltyConfig(): LoyaltyConfig {
  return { pointsName: '', subPoints: 0, resubPoints: 0, giftSubPoints: 0, cheerPointsPer100: 0, watchPointsPerTick: 0 };
}

// The broadcaster-facing counter scopes. 'command' pools every viewer into
// one total per command/reward; 'viewer_command' keeps one value per viewer
// per command. (A fifth, admin-only 'bot' scope exists service-side and never
// surfaces in the dashboard.)
export type CounterScope = 'channel' | 'viewer' | 'command' | 'viewer_command';
export const COUNTER_SCOPES: readonly CounterScope[] = ['channel', 'viewer', 'command', 'viewer_command'];

// CounterDef is one counter definition as counter.list returns it; value is
// the channel-scope tally (entry scopes keep per-viewer values instead).
export interface CounterDef {
  name: string;
  scope: CounterScope;
  value: number;
}

// CounterEntryView is one stored bucket of an entry-scoped counter (the
// per-counter leaderboard row).
export interface CounterEntryView {
  viewerId: string;
  viewerLogin: string;
  viewerName: string;
  command: string;
  value: number;
}

// LoyaltyStanding is one viewer's points + watch time (the channel top list).
export interface LoyaltyStanding {
  viewerId: string;
  viewerLogin: string;
  viewerName: string;
  points: number;
  watchSeconds: number;
}
